scripts/scheduler_ingest.py

```diff
@@ -0,0 +1,159 @@
+        """
+        Converts raw data received from the Scheduler API into a list of Cycle objects,
+        containing a list of MaintenanceDay objects - all which fall within the given cycle period.
+        :param raw_cycle_data: Cycle data received from the Scheduler API (as a list)
+        :param raw_maintenance_data: Maintenance day data received from the Scheduler API (as a list)
+        :return: A list of Cycle objects containing a list of 0 or more MaintenanceDay objects.
+        """
+        pre_processed = self._pre_process(raw_cycle_data, raw_maintenance_data)
+        return self._process(pre_processed)
+
+    @staticmethod
+    def print_start_dates(data):
+        """
+        Print each 'start' attribute within a list, prefixed by an index.
+        :param data: The list of items to print the start dates of
+        """
+        for idx, item in enumerate(data):
+            print(f"{idx}: {item['start']}")
+
+    def _pre_process(self, raw_cycle_data, raw_maintenance_data):
+        """
+        Makes adjustments to the data without converting it to a different data type.
+        These adjustments are necessary for the data to be properly processed
+        into Cycle and Maintenance Day objects.
+        :param raw_cycle_data: Cycle data received from the Scheduler API (as a list)
+        :param raw_maintenance_data: Maintenance day data received from the Scheduler API (as a list)
+        :return: A single list containing all valid items from both lists
+        """
+        data = self._combine_lists(raw_cycle_data, raw_maintenance_data)
+        data = self._sort_by_date(data)
+        data = self._clean_data(data)
+        return data
+
+    def _clean_data(self, data):
+        """
+        Removes invalid items from the data.
+        :param data: The list of items to be validated
+        :return: The list of items that were passed in, but with all invalid items removed
+        """
+        print("cleaning..")
+        print(f"Original length: {len(data)}")
+        clean_list = []
+        for item in data:
+            if 'name' in item and not re.search(self._cycle_name_regex, item['name']):
+                print(f"\nItem removed due to strange cycle name ({item['name']}).\nFull item: {item}")
+            elif item['start'].replace(tzinfo=None) < self._earliest_possible_date:
+                print(f"\nItem removed due to impossible date ({item['start']}).\nFull item: {item}")
+            else:
+                clean_list.append(item)
+        print(f"New length: {len(clean_list)}")
+        return clean_list
+
+    # TODO: question - I've used the method of combining the lists as this was suggested,
+    #  but I'm not exactly sure why we do this instead of keeping them separate?
+    #  Is it a speed optimisation perhaps? I wonder if it makes us more prone to bugs/confusing code..
+    @staticmethod
+    def _combine_lists(first, second):
```

Comment:
The reason to combine the lists is to make it easier/more efficient to process the maintenance days vs. the cycles. More specifically we start with:
```
maintenance_days = [suds.maintenance_day, ...]
cycles = [suds.cycles, ...]
```
from this point, we need to do the following steps to reach the desired output:
* Create Cycle / MaintenanceDay objects corresponding to each of the suds object
* Populate the Cycle objects with the maintenance days object that exist within their date range

By combining the lists first, we can sort both the cycles and maintenance_days by date. This means the list now looks like the following:
```
all_items = [cycle, maintenance_day, maintenance_day, cycle, maintenance_day, cycle, ...]
```
When the list is in this form it is easier and more efficient to convert the suds objects to our own custom data objects. The reason for this as we can loop through the full list and structure all the data in once because of the following assumption:
* In the sorted list, a cycle object is followed by the maintenance days that exist within its date range. 

Therefore we can create a cycle object then create maintenance day objects for all the following list items **until** we hit a new cycle object where we then can create a new cycle object and adding following maintenance days to it. 

tldr; it means that the complexity of this data structuring is `O(n)` as it's a single loop. If you store these lists separately, then its `O(n^2)` as you have to search through the maintenance day list every time you add the maintenance days to the cycles. It is marginally more efficient if both lists are sorted, but still not as fast as the first implementation.

Reply:
Having said all of this, we are talking about the efficiency over a rather small list (20+ items). I wonder if the readability would be better for separate lists where the process would be:
* sort both lists and keep them separate
* for each cycle:
  * create a new cycle object
  * look through the sorted maintenance day list
  * check to see if the date of the maintenance day falls within the cycle (should be the first n items as the list is sorted)
  * if maintenance day date is within cycle:
    * create maintenance day obj
    * add to cycle obj
    * remove maintenance suds obj from sorted list
  * else:
   * move to next cycle item

Reply:
okay before you do anything else can you implement the two sorted lists implementation and then we can compare the difference and decide which we want to go with? I agree that the current implementation is probably less obvious than using 2 lists.

Reply:
Very well summarised, thanks for this.
I see what you mean about keeping the complexity as o(n). What you've gone on to pseudocode there with the bullet points was actually what I was going to suggest! That method seems more readable, easier to add more helpful exceptions, and (if I add some well-placed conditions) should maintain o(n).

As you say, I'll implement the alternative and we'll go from there.

Reply:
*Interesting update:*
I've implemented your pseudo code (about to commit+push) and within it **included a check** that the latest **maintenance day ends before** the current **cycle's end date**.

On running - it's picked up an issue with the data I'm working with which the original method didn't pick up on
(see _output from the original_ method below with issue highlighted)

![image](https://user-images.githubusercontent.com/31534888/77647071-adfe7e80-6f5d-11ea-988f-6917fb6c0de4.png)

Reply:
Oh yes that's quite strange. Those maintenance days don't exist inside any cycle... is this for cycle 20_2. I wonder if it's an issue with the scheduling for it being off with all the current covid-19 stuff?

Reply:
Checking the beam status page, neither of the out of cycle maintenance day dates are on their table either. https://www.isis.stfc.ac.uk/Pages/Beam-Status.aspx
We should validate this with BusApps to see where these dates have come from. Although I suggest that this is low priority as the dates are for 20_1 which is not going to be going ahead.